lib/services/access_request_test.go

```diff
@@ -19,12 +19,630 @@ package services
+		"military": {
+			ReviewRequests: &types.AccessReviewConditions{
+				Roles: []string{"dictator"},
+				Where: `contains(request.system_annotations["mechanisms"],"coup") || contains(request.system_annotations["mechanism"],"treachery")`,
+			},
+		},
+	}
+
+	roles := make(map[string]types.Role)
+
+	for name, conditions := range roleDesc {
+		role, err := types.NewRole(name, types.RoleSpecV3{
+			Allow: conditions,
+		})
+		require.NoError(t, err)
+
+		roles[name] = role
+	}
+
+	// describes a collection of users with various roles
+	userDesc := map[string][]string{
+		"alice": {"populist", "proletariat", "intelligentsia", "military"},
+		"bob":   {"general", "proletariat", "intelligentsia", "military"},
+		"carol": {"conqueror", "proletariat", "intelligentsia", "military"},
+		"dave":  {"populist", "general", "conqueror"},
+		"erika": {"populist", "idealist"},
+	}
+
+	users := make(map[string]types.User)
+
+	for name, roles := range userDesc {
+		user, err := types.NewUser(name)
+		require.NoError(t, err)
+
+		user.SetRoles(roles)
+		users[name] = user
+	}
+
+	g := &mockUserAndRoleGetter{
+		roles: roles,
+		users: users,
+	}
+
+	const (
+		approve = types.RequestState_APPROVED
+		deny    = types.RequestState_DENIED
+	)
+
+	tts := []struct {
+		// requestor is the name of the requesting user
+		requestor string
+		// the roles to be requrested (defaults to "dictator")
+		roles   []string
+		reviews []struct {
+			// author is the name of the review author
+			author string
+			// noReview indicates that author will not be allowed to review
+			noReview bool
+			// propose is the state proposed by the review
+			propose types.RequestState
+			// expect is the expected post-review state of the request (defaults to pending)
+			expect types.RequestState
+		}
+	}{
+		{
+			requestor: "alice", // permitted by role populist
+			reviews: []struct {
+				author          string
+				noReview        bool
+				propose, expect types.RequestState
+			}{
+				{ // cannot review own requests
+					author:   "alice",
+					noReview: true,
+				},
+				{ // no matching allow directives
+					author:   g.user(t, "military"),
+					noReview: true,
+				},
+				{ // adds one approval to all thresholds
+					author:  g.user(t, "proletariat", "intelligentsia", "military"),
+					propose: approve,
+				},
+				{ // adds one denial to all thresholds
+					author:  g.user(t, "proletariat", "intelligentsia", "military"),
+					propose: deny,
+				},
+				{ // adds second approval to all thresholds, triggers "uprising".
+					author:  g.user(t, "proletariat", "intelligentsia", "military"),
+					propose: approve,
+					expect:  approve,
+				},
+				{ // adds second denial to all thresholds, no effect since a state-transition was already triggered.
+					author:  g.user(t, "proletariat", "intelligentsia", "military"),
+					propose: deny,
+					expect:  approve,
+				},
+			},
+		},
+		{
+			requestor: "alice", // permitted by role populist
```

Comment:
nit: what's the reason to split these reviews into two groups for `alice`?

Reply:
Testing two separate scenarios is all.  Since each set of reviews are applied to the same underlying request, we need to split this scenario off into a separate block in order to start with a clean request state.